develop/tutorials/articles/03-developing-a-web-application/04-writing-web-application-liferay/01-writing-a-web-application.markdown

```diff
@@ -0,0 +1,635 @@
+best way to implement a mechanism for saving guestbook entries. This approach
+is easy to implement but it's really just a shortcut that should only be
+considered when developing a prototype application. To learn how to implement a
+robust mechanism for saving application data, follow the *Writing a Data-Driven
+Application* learning path.
+
+$$$
+
+The following method implements adding a guestbook entry to a portlet preference
+called `guestbook-entries`:
+
+    public void addEntry(ActionRequest request, ActionResponse response) {
+        try {
+            PortletPreferences prefs = request.getPreferences();
+
+            String[] guestbookEntries = prefs.getValues("guestbook-entries",
+                    new String[1]);
+
+            ArrayList<String> entries = new ArrayList<String>();
+
+            if (guestbookEntries != null) {
+                entries = new ArrayList<String>(Arrays.asList(prefs.getValues(
+                        "guestbook-entries", new String[1])));
+            }
+
+            String userName = ParamUtil.getString(request, "name");
+            String message = ParamUtil.getString(request, "message");
+            String entry = userName + "^" + message;
+
+            entries.add(entry);
+
+            String[] array = entries.toArray(new String[entries.size()]);
+
+            prefs.setValues("guestbook-entries", array);
+
+            try {
+                prefs.store();
+            }
+            catch (IOException ex) {
+                Logger.getLogger(GuestbookPortlet.class.getName()).log(
+                        Level.SEVERE, null, ex);
+            }
+            catch (ValidatorException ex) {
+                Logger.getLogger(GuestbookPortlet.class.getName()).log(
+                        Level.SEVERE, null, ex);
+            }
+
+        }
+        catch (ReadOnlyException ex) {
+            Logger.getLogger(GuestbookPortlet.class.getName()).log(
+                    Level.SEVERE, null, ex);
+        }
+    }
+
+First, the preferences are retrieved, and then the `guestbook-entries`
+preference is retrieved and converted to an `ArrayList` so that you can add an
+entry without worrying about exceeding the size of the array. Next, the name
+and message fields from your form are retrieved. Note that Liferay's
+`ParamUtil` class makes it very easy to retrieve URL parameters.
+
+Finally, the fields are combined into a `String` delimited by a caret, and the
+new entry is added to the `ArrayList`, which is then converted back to an array
+so it can be stored as a preference. The try/catch blocks are required by the
+portlet preferences API.
+
+This is not the normal way to use portlet preferences, but it provides a quick
+and easy way for you to store guestbook entries in this first version of your
+application. You'll implement a robust way to store guestbook entries in a
+later learning path.
+
+The next and final feature to implement is a mechanism for viewing guestbook
+entries.
+
```

Comment:
Probably break it up again here. Again, we're looking for short, sweet, bite-size steps.